### PR TITLE
WIP: SilverStripe4 compatibility

### PR DESCRIPTION
**Pull request by stephenmcm**
This PR is partially for discussion as it's slightly easier to track what's changed than flipping between repositories. In it's current state it is stable and used in production but further improvements are always possible.

## Changelog
 
 - Moved code under src/
 - Namespaced everything under _2fa
 - Moved yml values around so less sits under Member
 - Moved as much of the token logic out of Member and into Authenticator as was straight forward

### Added

 - Authenticator and LoginHandler following bootstrapMFA's pattern
 - SiteConfigExtension allows CMS based enable/disable and enforce. Over-ridable via YML
 - Added basic set up of 2fa on the front end. This needs some further extension, but for my current use case it works well as enforcing 2fa
 - Add some basic tests

### Removed

 - LoginForm
 - migration/BackupToken

### CODE REVIEW

**Comment by oddnoc** [comment]
Hi Stephen & Cam,

Thanks for providing this, Stephen. I've started testing and found some issues.

1. (minor) It took a few tries before backup codes appeared.
1. ~(trivial) Selecting the backup codes also selects an invisible "Save" string.~
1. (major) `_2fa\Authenticator\is2FAenabled()` has a dependency on SiteConfig, which breaks sites that don't include the silverstripe/siteconfig module.

Related to this last item: I think probably I ought not to have been able to enable 2fa on my account without first having set `enable_2fa` in a configuration file.

**Comment by camfindlay** [comment]
Happy for you two to proceed as you see fit on this PR :+1:

**Comment by oddnoc** [comment]
I've added 2 commits:

1. 4b73c3a8736ab801390f9c2119d5060d1133a34b -- adds siteconfig dependency
1. 7b6286a74881fad54151dcaaf2840d5642cec638 -- fixes a typo

The problem with the "Save" string getting selected was my fault (I triple-clicked to start my selection).

**Comment by stephenmcm** [comment]
Hi Fred,

Yeah good point on being able to turn on 2fa without it enabled. A quick check for the setting in the CMSProfileController will fix that. I've moved the is2FAenabled and is2FArequired functions into the Authenticator, as I believe this is a reasonable place for them. And added a if is2FAenabled check in the CMSProfileController, whcih I'll push up soon.

The back up codes are indeed buggy right now, I'll need to check the logic to work out why.

For the siteconfig dependency we could move the config to a dedicated model admin but I don't think that's a great user experience.

**Comment by oddnoc** [comment]
I've just pushed 2e49ffc to remove the 2FA UI when 2FA is not enabled.

**Comment by stephenmcm** [comment]
Looks like I need to include those commits as patches? Git doesn't give me an option to pull them down. Either way the 4b73c3a and 7b6286a will be fine. I've overridden the 2e49ffc commit as it needs to check for either the SiteConfig or yaml value.

I've also set the back up tokens to regenerate when the user activates 2FA, this should resolve issue 1.

**Comment by camfindlay** [comment]
Sign out when you need some user testing on this, happy to run through use from an end user pov to see if it's all working as expected.

**Comment by stephenmcm** [comment]
Hey Cam, I'm reasonably happy with where this is now. There's a couple of things that need documentation, but the usability is all there. I'm also not a huge fan of the namespacing choices but that's a nitpick rather than a true issue. Feel free to give it a UX level test now, or let me know what you expect to work and I'll check that it's functional.

**Comment by camfindlay** [comment]
What would you do different with the name spaces? If we're going to release a major version with breaking changes... now's the time to adjust ;)

File: tests/CMSProfileControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        $controller->setRequest(Controller::curr()->getRequest());
+        $form = $controller->getEditForm($member->ID);
+        $this->assertNull($form->Fields()->fieldByName('Root.TwoFactorAuthentication.BackupTokens'));
+        $this->assertInstanceOf(FormAction::class, $form->Fields()->fieldByName('Root.TwoFactorAuthentication.action_open_deactivation_dialog'));
+    }
+
+    public function testActivate2FA()
+    {
+        $this->markTestIncomplete('New activation path requires confirm 2FA token');
+        // $member = $this->objFromFixture(Member::class, 'user1');
```
Comment (camfindlay) [inline tests/CMSProfileControllerTest.php:56]: worth tidy up on comments code?

**Comment by oddnoc** [comment]
It looks like Github is getting in the way of us merging our work together. I've pushed my changes to feature/ss4 on https://github.com/Quinn-Interactive/silverstripe-twofactorauth. It is rebased on @stephenmcm's feature-ss4 branch.

**Comment by camfindlay** [comment]
Looks like it's just the composer.json that is in conflict.... can either of you resolve locally and push up a resolved composer.json... should solve the issue.

Also suggest that we squash commits before merging.

File: tests/CMSProfileControllerTest.php (hunk 1 above)
Comment (stephenmcm) [reply tests/CMSProfileControllerTest.php:56]: Yeah we should. I'll give the tests section a once over to see if there's any easy wins for more tests.

**Comment by stephenmcm** [comment]
Sorry for the slow reply, I was off on a short holiday.
Personally, I dislike the `_2fa` thing and would far prefer just `twofactorauth` to be 1-to-1 with the module name. I also wonder if perhaps we should make the namespace paths of overriding classes match the SilverStripe core class e.g:
```
  SilverStripe\Security\ChangePasswordForm:
    class: _2fa\ChangePasswordForm
```
To:
```
  SilverStripe\Security\ChangePasswordForm:
    class: _2fa\Security\TwoFactorAuthChangePasswordForm
```

I'll try to get to fixing the composer.json conflict this week. And I strongly suggest we don't squash the commits, we've just had an issue locally were a large squash merge has caused lost history making a bug harder to track down. Though a clean up with a `rebase -i` might be worthwhile to clean up any typos and or 1 line change commits.

**Comment by camfindlay** [comment]
hi @stephenmcm @oddnoc  - I'm working with @wilr on a project and he's able to look at getting this PR tidied up and merged. I'm going to add him as a maintainer just an FYI.

**Comment by wilr** [comment]
I've fixed the merge conflict. LGTM!

**Pull request merged**